### PR TITLE
Homework - Cars

### PR DESCRIPTION
**Pull request by Pawellos**
- Added Exceptions 
- Improved encapsulation 
- Added setters/getters
- Fixed memory leaks - Using smart pointers 
- Added Tests

### CODE REVIEW

File: PetrolEngine.cpp [hunk 1]
```diff
@@ -2,19 +2,36 @@
...
-{
-    // TODO: Add checking if gear is between -1 and gears_
-    // -1 is for REAR
-    // 0 is for NEUTRAL
+PetrolEngine::~PetrolEngine() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void PetrolEngine::changeGear(int gear) {
+    if (gear < -1 || gear > gears_) {
```
Comment (MaciejSalamonski) [inline PetrolEngine.cpp:14]: O think that You should avoid magic numbers. Let's make - 1 as constexpr or const variable

**Review by MaciejSalamonski — APPROVED** [review]
Very good code. Have a great day!

File: PetrolEngine.cpp (hunk 1 above)
Comment (WBigaj) [reply PetrolEngine.cpp:14]: +1

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: Car.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+}
+void Car::accelerate(int speed) {
+    if (speed < 0) {
+        throw InvalidSpeed("Can not drive with speed less than 0");
+    } else {
+        speed_ = speed;
+    }
+}
+
+void Car::showSpeed() {
```
Comment (niemarcin) [inline Car.cpp:22]: Might be marked as const method.

File: Car.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    std::cout << "Speed = " << speed_ << '\n';
+}
+
+int Car::getSpeed() {
```
Comment (niemarcin) [inline Car.cpp:26]: Might be marked as const method.

File: Car.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+#include "Car.hpp"
+#include "iostream"
```
Comment (niemarcin) [inline Car.cpp:2]: Wouldn't be better to write <iostream> rather than "iostream" to mark it's standard library?

File: ElectricCar.cpp [hunk 5]
```diff
@@ -1,16 +1,29 @@
...
+
+void ElectricCar::changeEngine(std::unique_ptr<ElectricEngine> engine) {
+    if (speed_ == 0) {
+        engine_ = std::move(engine);
+    } else {
+        throw InvalidEngine("Can not change the engine during a drive !");
+    }
+}
 
+void ElectricCar::getInfoEngine() {
```
Comment (niemarcin) [inline ElectricCar.cpp:27]: Might be marked as const method.

File: Car.hpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+public:
+    virtual ~Car();
+    void turnLeft();
+    void turnRight();
+    void brake();
+    void accelerate(int speed);
+    void showSpeed();
+    int getSpeed();
+    virtual void restore() = 0;
+    virtual void getInfoEngine() = 0;
```
Comment (niemarcin) [inline Car.hpp:16]: Might be marked as const method.
And maybe "print" would be better than "get" - it's strange that "get" method does not return anything.

File: PetrolEngine.cpp [hunk 7]
```diff
@@ -2,19 +2,36 @@
...
+    }
     currentGear_ = gear;
     std::cout << __FUNCTION__ << std::endl;
 }
+
+void PetrolEngine::showCurrentGear() {
+    std::cout << "Current gear = " << currentGear_ << "\n";
+}
+
+int PetrolEngine::getCurrentGear() {
```
Comment (niemarcin) [inline PetrolEngine.cpp:27]: Might be marked as const method.

File: PetrolEngine.cpp [hunk 8]
```diff
@@ -2,19 +2,36 @@
...
+    if (gear < -1 || gear > gears_) {
+        throw InvalidGear("Wrong gear");
+    } else if (abs(gear - currentGear_) > 2) {
+        throw InvalidGear("Not allowed gears changing more two");
+    }
     currentGear_ = gear;
     std::cout << __FUNCTION__ << std::endl;
 }
+
+void PetrolEngine::showCurrentGear() {
```
Comment (niemarcin) [inline PetrolEngine.cpp:23]: Might be marked as const method.

**Review by niemarcin — APPROVED** [review]
Multiple methods might be marked as const method. Not a big deal but worth to do.

**Review by jakjan95 — APPROVED** [review]
Code looks good :) If you have time try to add more improvements

File: HybridCar.cpp [hunk 9]
```diff
@@ -1,19 +1,29 @@
...
+    PetrolCar::printInfoEngine();
+    ElectricCar::printInfoEngine();
+}
+
+void HybridCar::changePetrolEngine(std::unique_ptr<PetrolEngine> petrolEngine) {
+    PetrolCar::changeEngine(std::move(petrolEngine));
+}
 
+void HybridCar::changeElectricEngine(std::unique_ptr<ElectricEngine> electricEngine) {
+    ElectricCar::changeEngine(std::move(electricEngine));
```
Comment (ziobron) [inline HybridCar.cpp:28]: redundant functions. We inherit changeEngine from base classes, so we can call changeEngine with either unique_ptr to Petrol or Electric engine.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: ElectricCar.cpp [hunk 10]
```diff
@@ -1,16 +1,29 @@
...
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void ElectricCar::restore() {
+    charge();
+}
+
+void ElectricCar::changeEngine(std::unique_ptr<ElectricEngine> engine) {
+    if (speed_ == 0) {
+        engine_ = std::move(engine);
```
[the remark this answers is not in the document]
Comment (Pawellos) [reply ElectricCar.cpp]: I thought the smart pointers close the deal :)

[the remark this answers is not in the document]
Comment (ziobron) [reply ElectricCar.cpp]: Heh, you are right. My bad. I confused this homework with the one about unique_ptrs, where you have had to implement your own move semantics :)

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
5 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119] }

**Pull request closed without merging**